src/Sampler.cpp

```diff
@@ -0,0 +1,233 @@
+                              comm, vertex_edge, one_weight, W_block, partitioning, edge_d_td,
+                              edge_boundary_att, ess_attr, spect_tol, max_evects, dual_target,
+                              scaled_dual, energy_dual, hybridization);
+    fvupscale_temp->MakeFineSolver();
+    fvupscale_ = fvupscale_temp;
+
+    Initialize(dimension, kappa);
+}
+
+void PDESampler::Initialize(int dimension, double kappa)
+{
+    rhs_fine_.SetSize(fine_vector_size_);
+    coefficient_fine_.SetSize(fine_vector_size_);
+    rhs_coarse_ = fvupscale_->GetCoarseVector();
+    coefficient_coarse_.SetSize(num_coarse_aggs_);
+
+    double nu_parameter;
+    MFEM_ASSERT(dimension == 2 || dimension == 3, "Invalid dimension!");
+    if (dimension == 2)
+        nu_parameter = 1.0;
+    else
+        nu_parameter = 0.5;
+    double ddim = static_cast<double>(dimension);
+    scalar_g_ = std::pow(4.0 * M_PI, ddim / 4.0) * std::pow(kappa, nu_parameter) *
+                std::sqrt( tgamma(nu_parameter + ddim / 2.0) / tgamma(nu_parameter) );
+}
+
+PDESampler::~PDESampler()
+{
+}
+
+/// @todo cell_volume should be variable rather than constant
+void PDESampler::NewSample()
+{
+    current_state_ = FINE_SAMPLE;
+
+    // construct white noise right-hand side
+    // (cell_volume is supposed to represent fine-grid W_h)
+    for (int i = 0; i < fine_vector_size_; ++i)
+    {
+        rhs_fine_(i) = scalar_g_ * std::sqrt(cell_volume_) *
+                       normal_distribution_.Sample();
+    }
+}
+
+void PDESampler::NewCoarseSample()
+{
+    current_state_ = COARSE_SAMPLE;
+    MFEM_ASSERT(false, "Not implemented!");
+}
+
+mfem::Vector& PDESampler::GetFineCoefficient()
+{
+    MFEM_ASSERT(current_state_ == FINE_SAMPLE,
+                "PDESampler object in wrong state (call NewSample() first)!");
+
+    fvupscale_->SolveFine(rhs_fine_, coefficient_fine_);
+    for (int i = 0; i < coefficient_fine_.Size(); ++i)
+    {
+        coefficient_fine_(i) = std::exp(coefficient_fine_(i));
+    }
+    return coefficient_fine_;
+}
+
+/**
+   For myself:
+
+   c_i              : coefficient for coarse basis function, representing ~normal field K
+   (c_i / q_i)      : value of ~normal field K on agg i
+   exp(c_i/q_i)     : value of lognormal field exp(K) on agg i
+   exp(c_i/q_i) q_i : coefficient for coarse basis function, representing lognormal field exp(K)
+
+   indexing: we consider only the coefficient for the *constant* component i
+*/
+mfem::Vector& PDESampler::GetCoarseCoefficient()
+{
+    MFEM_ASSERT(current_state_ == FINE_SAMPLE ||
+                current_state_ == COARSE_SAMPLE,
+                "PDESampler object in wrong state (call NewSample() first)!");
+
+    if (current_state_ == FINE_SAMPLE)
+        fvupscale_->Restrict(rhs_fine_, rhs_coarse_);
+    mfem::Vector coarse_sol = fvupscale_->GetCoarseVector();
+    fvupscale_->SolveCoarse(rhs_coarse_, coarse_sol);
+    coarse_sol *= -1.0; // ?
```

Comment:
Compare with `GetFineCoefficient`, the solution here is not orthogonalized to constants. I dont know if that is necessary, I just notice the difference between the two (and we have the tool to do the average).